File: java/day03/06. 两两交换链表中的节点.java
/**
 * https://leetcode.cn/problems/swap-nodes-in-pairs/
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode x = head;
        ListNode y = head.next;
        ListNode z = head;
        head = head.next;
        while (x != null && y != null) {
            z.next = y;
            x.next = y.next;
            y.next = x;
            z = x;
            x = x.next;
            if (x != null) {
                y = x.next;
            }
        }
        return head;
    }
}

File: ts/06.两两交换链表中的节点.ts
function swapPairs(head: ListNode | null): ListNode | null {
  // 如果当前节点或者下一个节点为null，则直接返回当前节点
  if (head === null || head.next === null) {
    return head;
  }

  const next = head.next;

  head.next = swapPairs(next.next);
  next.next = head;

  return next;
}
